### PR TITLE
Update abci/app docs

### PR DESCRIPTION
**Pull request by ebuchman**
Improve documentation, including what's noted in https://github.com/tendermint/tendermint/pull/2467#issuecomment-423682294 and various issues.

### CODE REVIEW

**Comment by ebuchman** [comment]
Merging, but please still review :)

File: docs/spec/abci/apps.md [hunk 1]
```diff
@@ -86,18 +86,50 @@ Otherwise it should never be modified.
...
-though see issues
-[#1861](https://github.com/tendermint/tendermint/issues/1861),
-[#2299](https://github.com/tendermint/tendermint/issues/2299) and
-[#2310](https://github.com/tendermint/tendermint/issues/2310) for how this may
-soon change.
+The `Data` field must be strictly deterministic, but can be arbitrary data.
+
+### Gas
+
+Ethereum introduced the notion of `gas` as an absract representation of the
```
Comment (melekes) [inline docs/spec/abci/apps.md]: abstract

File: docs/spec/abci/apps.md [hunk 2]
```diff
@@ -86,18 +86,50 @@ Otherwise it should never be modified.
...
+    - `GasWanted <= MaxGas` for all txs in the mempool
+    - `(sum of GasWanted in a block) <= MaxGas` when proposing a block
+
+If `MaxGas == -1`, no rules about gas are enforced.
+
+Note that Tendermint does not currently enforce anything about Gas in the consensus, only the mempool.
+This means it does not guarantee that committed blocks satisfy these rules!
+It is the application's responsibility to return non-zero response codes when gas limits are exceeded.
+
+The `GasUsed` field is ignored compltely by Tendermint. That said, applications should enforce:
```
Comment (melekes) [inline docs/spec/abci/apps.md]: completely

File: docs/tendermint-core/running-in-production.md [hunk 3]
```diff
@@ -1,5 +1,41 @@
 # Running in production
 
+## Database
+
+By default, Tendermint uses the `syndtr/goleveldb` package for it's in-process
+key-value database. Unfortunately, this implementation of LevelDB seems to suffer under heavy load (see
+[#226](https://github.com/syndtr/goleveldb/issues/226)). It may be best to
+install the real C-implementaiton of LevelDB and compile Tendermint to use
```
Comment (melekes) [inline docs/tendermint-core/running-in-production.md]: implementation

**Pull request merged**